src/libstd/keyword_docs.rs

```diff
@@ -1578,12 +1578,108 @@ mod type_keyword {}
 
 #[doc(keyword = "unsafe")]
 //
-/// Code or interfaces whose [memory safety] cannot be verified by the type system.
+/// Code or interfaces whose [memory safety] cannot be verified by the type
+/// system.
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
+/// The `unsafe` keyword has two uses: to declare the existence of contracts the
+/// compiler can't check, and to declare that a programmer has checked that
+/// these contracts have been upheld.
+///
+/// Unsafe Rust add some capabilities to Safe Rust:
+///
+/// - Dereference [raw pointers]
+/// - Call `unsafe` functions (including foreign functions, compiler intrinsics,
+/// and the raw allocator)
+/// - Implement `unsafe` [`trait`]s
+/// - Mutate [`static`]s (including [`extern`]al ones)
+/// - Access fields of [`union`]s
+///
+/// # Dereferencing raw pointers
+///
+/// ```rust
+/// let a = 3;
+/// let b = &a as *const _;
+/// // SAFETY: we know that b will always be a valid address.
+/// let _ = unsafe { *b };
+/// ```
+///
+/// # Marking elements as unsafe
+///
+/// `unsafe` can be used on functions. Note that functions declared in
+/// [`extern`] blocks are implicitly marked as `unsafe`. Methods can also be
+/// declared as `unsafe`:
+///
+/// ```rust
+/// #![allow(dead_code)]
+/// unsafe fn unsafe_fn() {}
+///
+/// extern "C" {
+///     fn unsafe_extern_fn();
+/// }
+///
+/// trait TraitWithUnsafeMethod {
+///     unsafe fn unsafe_method();
+/// }
+///
+/// struct S;
+///
+/// impl S {
+///     unsafe fn unsafe_method_on_struct() {}
+/// }
+/// ```
+///
+/// Traits can also be declared as `unsafe`:
```

Comment:
I think it would be good to either link to the discussion in https://internals.rust-lang.org/t/what-does-unsafe-mean/6696 or summarize it here. tl;dr, `unsafe fn`, `unsafe {}`, and `unsafe impl` mean very different things.

Reply:
I don't understand unsafe traits reading this, maybe have a good example of why do we need unsafe traits?

Reply:
https://doc.rust-lang.org/nomicon/send-and-sync.html

> Some types allow you to have multiple aliases of a location in memory while mutating it. Unless these types use synchronization to manage this access, they are absolutely not thread-safe. Rust captures this through the Send and Sync traits.

>    A type is Send if it is safe to send it to another thread.
>    A type is Sync if it is safe to share between threads (&T is Send).

These are automatically derived, but they are not part of the language. Additionally, adding `impl<T> Sync for Rc<T>` would be UB because it could lead to double-frees (e.g. if the counter is not updated atomically).

I recommended a short summary of this in https://github.com/rust-lang/rust/pull/73943#discussion_r448649436 but it's not a bad idea to go more in depth.

Reply:
I don't know if keyword doc is the best place to go in depth about that. I'll add links in a section about the different meaning of `unsafe` for now

Reply:
Sure, links are fine. I just think it's important to emphasize that it means different things in different contexts.